exporters/dynamicconfig/confignotifier.go

```diff
@@ -0,0 +1,207 @@
+
+	return configString == otherConfigString
+}
+
+// TODO: Read actual config from host
+// We don't do anything with the configHost string right now
+// TIMES_READ_TEST is for testing purposes, to mimic a dynamic service
+// The second time we call readConfig, the sampling period changes
+var TIMES_CONFIG_READ_TEST = 0
+
+func readConfig(configHost string) *MetricConfig {
+	period := 10 * time.Second
+
+	if TIMES_CONFIG_READ_TEST > 0 {
+		period = 2 * time.Second
+	}
+
+	TIMES_CONFIG_READ_TEST++
+
+	return &MetricConfig{
+		Period: period,
+	}
+}
+
+type ConfigWatcher interface {
+	OnInitialConfig(config *MetricConfig)
+	OnUpdatedConfig(config *MetricConfig)
+}
+
+// A ConfigNotifier monitors a config service for a config changing
+// It then lets all it's subscribers know if the config has changed
+type ConfigNotifier struct {
+	// Used to shut down the config checking routine when we stop ConfigNotifier
+	ch chan struct{}
+
+	// How often we check to see if the config service has changed
+	checkFrequency time.Duration
+
+	// Added for testing time-related functionality
+	clock controllerTime.Clock
+
+	// Current config
+	config *MetricConfig
+
+	// Optional if config is non-dynamic. Address of config service host
+	configHost string
+
+	lock sync.Mutex
+
+	// Contains all the notifier's subscribers
+	subscribed map[ConfigWatcher]bool
+
+	// Controls when we check the config service for a new config
+	ticker controllerTime.Ticker
+
+	// Used to wait for the config checking routine to return when we stop the notifier
+	wg sync.WaitGroup
+}
+
+// Constructor for a ConfigNotifier
+func New(checkFrequency time.Duration, defaultConfig *MetricConfig, opts ...Option) *ConfigNotifier {
+	notifier := &ConfigNotifier{
+		ch:             make(chan struct{}),
+		checkFrequency: checkFrequency,
+		clock:          controllerTime.RealClock{},
+		config:         defaultConfig,
+		subscribed:     make(map[ConfigWatcher]bool),
+	}
+
+	for _, opt := range opts {
+		opt.Apply(notifier)
+	}
+
+	return notifier
+}
+
+// SetClock supports setting a mock clock for testing.  This must be
+// called before Start().
+func (n *ConfigNotifier) SetClock(clock controllerTime.Clock) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	n.clock = clock
+}
+
+func (n *ConfigNotifier) Start() {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	if n.configHost == "" {
+		return
+	}
+
+	if n.ticker != nil {
+		return
+	}
+
+	n.ticker = n.clock.Ticker(n.checkFrequency)
+	n.wg.Add(1)
+	go n.checkChanges(n.ch)
+}
+
+func (n *ConfigNotifier) Stop() {
+	n.lock.Lock()
+
+	if n.configHost == "" {
+		return
+	}
+
+	if n.ch == nil {
+		return
+	}
+	close(n.ch)
+	n.ch = nil
+
+	n.lock.Unlock()
+
+	n.wg.Wait()
+	n.ticker.Stop()
+}
+
+func (n *ConfigNotifier) Register(watcher ConfigWatcher) {
+	n.lock.Lock()
+	n.subscribed[watcher] = true
+
+	// Avoid lock starvation with OnInitialConfig by making a copy of the config
+	config_copy := *n.config
+	n.lock.Unlock()
+
+	watcher.OnInitialConfig(&config_copy)
```

Comment:
Thinking more about this I think it may be problematic if we do this like this. The following sequence of events is possible (in chronicle order):

1. Watcher (W1) register an initial config handle.
2. The watcher W1 is added, but "watcher.OnInitialConfig" is not called because the OS decides to schedule something else.
3. A new config config come.
4. The new config is applied to W1 in checkChanges() with watcher.OnUpdatedConfig().
5. The Register() is re-scheduled by the OS and resumes its execution, so "watcher.OnInitialConfig(&config_copy)" is called.

So, we end up overwriting the new updated config and incorrectly using the old initial config as the most up-to-date config.